API-GUIDELINES.md

```diff
@@ -0,0 +1,54 @@
+# ESP-RS API Guidelines
+
+## About
+
+This is a living document - make sure to check the latest version of this document.
+
+> [!NOTE]
+> Not all of the currently existing code follows this guideline, yet.
+
+In general, the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines) apply to all projects in the ESP-RS GitHub organization where possible. (e.g. `C-SERDE` does not apply) 
+
+Especially for public API but if possible also for internal APIs.
+
+The following paragraphs contain additional recommendations.
+
+## Construction and Destruction of Drivers
+
+- Drivers take peripherals and pins via the `PeripheralRef` pattern - they don't consume peripherals/pins
+- Consider adding a `Drop` implementation resetting the peripheral to idle state
+- Consider using a builder-like pattern for configuration which must be done during initialization
+
+## Interoperability
+
+- `cfg` gated `defmt` derives and impls are added to new structs and enums
+    - see [this example](https://github.com/esp-rs/esp-hal/blob/df2b7bd8472cc1d18db0d9441156575570f59bb3/esp-hal/src/spi/mod.rs#L15)
+    - e.g. `#[cfg_attr(feature = "defmt", derive(defmt::Format))]`
+- Don't use `log::XXX!` macros directly - use the wrappers in `fmt.rs` (e.g. just `info!` instead of `log::info!` or importing `log::*`)
+
+## API Surface
+
+- Add `#[deny(missing_docs)]` to new modules or when reworking a larger part of a module. In the end we will require this for whole crates.
+- API documentation shouldn't be an afterthought
+- Private details shouldn't leak into the public API, and should be made private where technically possible.
+- Implementation details that _need_ to be public should be marked with `#[doc(hidden)]` and a comment as to why it needs to be public.
+- Functions which technically need to be public but shouldn't be callable by the user need to be sealed.
+    - see [this example in Rust's core library](https://github.com/rust-lang/rust/blob/044a28a4091f2e1a5883f7fa990223f8b200a2cd/library/core/src/error.rs#L89-L100)
+- Any public traits, that **must not** be implemented downstream need to be `Sealed`
+- Prefer compile-time checks over runtime checks where possible, prefer a fallible API over panics.
+- Follow naming conventions in order to be consistent across drivers - take inspiration from existing drivers
```

Comment:
Maybe we can link here an example of "good" driver i.e some of our drivers need big refactoring so being explicit here might be helpful.

Reply:
Yes, that would be good - but I struggled choosing one. Some are in good shape but are relatively simple and some are quite complex. Open for suggestions here

Reply:
Maybe we can link both, "simple" and "complex"? IMO it won't bring too much noise here.